### PR TITLE
docs(ast): add doc comments to more AST nodes

### PR DESCRIPTION
**Pull request by DonIsaac**
# What This PR Does
Adds more doc comments and examples to AST nodes

### CODE REVIEW

File: crates/oxc_ast/src/ast/js.rs [hunk 1]
```diff
@@ -171,34 +173,57 @@ pub struct IdentifierName<'a> {
...
+    ///
+    /// Identifies what identifier this refers to, and how it is used. This is
+    /// set in the bind step of semantic analysis, and will always be [`None`]
+    /// immediately after parsing.
     #[serde(skip)]
     pub reference_id: Cell<Option<ReferenceId>>,
+    /// Flags indicating how the reference is used.
+    ///
+    /// This gets set in the bind step of semantic analysis, and will always be
+    /// [`ReferenceFlag::empty`] immediately after parsing.
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast/js.rs:197]: I think this should be `ReferenceFlag::None`.

File: crates/oxc_ast/src/ast/js.rs [hunk 2]
```diff
@@ -465,6 +535,15 @@ pub struct CallExpression<'a> {
...
 
 /// New Expression
+///
+/// ## Example
+/// ```ts
+/// //           callee         arguments
+/// //              ↓↓↓         ↓↓↓↓         
+/// const foo = new Foo<number>(1, 2)
+/// //                 ↑↑↑↑↑↑↑↑
+/// //                 type parameters
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast/js.rs:545]: `type_parameters`

File: crates/oxc_ast/src/ast/js.rs [hunk 3]
```diff
@@ -283,7 +317,22 @@ pub struct ObjectExpression<'a> {
 #[cfg_attr(feature = "serialize", derive(Serialize, Tsify))]
 #[serde(untagged)]
 pub enum ObjectPropertyKind<'a> {
+    /// Object Property
+    ///
+    /// ## Example
+    /// ```ts
+    /// const foo = 'foo'
+    /// const x = { foo, bar: 'bar' }
+    /// //          ^^^  ^^^^^^^^^
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast/js.rs:326]: Add one more `^` on end to go under second `'`.

File: crates/oxc_ast/src/ast/js.rs [hunk 4]
```diff
@@ -451,6 +508,19 @@ pub struct PrivateFieldExpression<'a> {
...
+/// ```ts
+/// //        ___ callee
+/// const x = foo(1, 2)
+///
+/// //            ^^^^ arguments
+/// const y = foo.bar?.(1, 2)
+/// //               ^ optional
+///
+/// const z = foo<number, string>(1, 2)
+/// //            ^^^^^^^^^^^^^^ type parameters
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast/js.rs:522]: `type_parameters`

File: crates/oxc_ast/src/ast/js.rs [hunk 5]
```diff
@@ -451,6 +508,19 @@ pub struct PrivateFieldExpression<'a> {
 }
 
 /// Call Expression
+///
+/// ## Examples
+/// ```ts
+/// //        ___ callee
+/// const x = foo(1, 2)
+///
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast/js.rs:516]: Don't think this line break should be here.

File: crates/oxc_ast/src/ast/js.rs [hunk 6]
```diff
@@ -1610,15 +1757,45 @@ pub struct ClassBody<'a> {
...
+/// ## Example
+/// ```ts
+/// class Foo {
+///   [prop: string]: string // ClassElement::TSIndexSignature
+///
+///   public x: number // ClassElement::PropertyDefinition
+///
+///   accessor z() { return 5 } // ClassElement::AccessorProperty
+///
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast/js.rs:1770]: Remove one line break.

File: crates/oxc_ast/src/ast/js.rs [hunk 7]
```diff
@@ -1828,22 +2087,45 @@ pub struct ImportSpecifier<'a> {
...
+    /// The name of the imported symbol.
     pub local: BindingIdentifier<'a>,
 }
 
-// import * as local from "source"
+/// Namespace import specifier
+///
+/// ## Example
+/// ```ts
+/// import * as local from "source"
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast/js.rs:2127]: Add semicolon on end (for consistency).

File: crates/oxc_ast/src/ast/ts.rs [hunk 8]
```diff
@@ -1174,12 +1210,15 @@ pub struct TSInstantiationExpression<'a> {
     pub type_parameters: Box<'a, TSTypeParameterInstantiation<'a>>,
 }
 
+/// See [TypeScript - Type-Only Imports and Exports](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html)
 #[ast]
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
 #[cfg_attr(feature = "serialize", derive(Serialize, Tsify))]
 #[serde(rename_all = "camelCase")]
 pub enum ImportOrExportKind {
+    /// `import { foo } from './foo'`
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast/ts.rs:1219]: Add semicolon on end (for consistency). Ditto 2 lines later.

File: crates/oxc_ast/src/ast_impl/js.rs [hunk 9]
```diff
@@ -291,6 +299,12 @@ impl<'a> Hash for IdentifierReference<'a> {
     }
 }
 
+impl<'a> PartialEq<IdentifierReference<'a>> for IdentifierReference<'a> {
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast_impl/js.rs:302]: Please do this in a separate PR (and same for other `impl`s added).

File: crates/oxc_ast/src/ast_impl/js.rs [hunk 10]
```diff
@@ -345,7 +375,7 @@ impl<'a> PropertyKey<'a> {
                 .is_empty()
                 .then(|| lit.quasi())
                 .flatten()
-                .map(std::convert::Into::into),
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast_impl/js.rs:348]: This is a good change, but please move it into a separate PR.

File: crates/oxc_ast/src/ast_impl/js.rs [hunk 11]
```diff
@@ -1306,28 +1381,32 @@ impl PropertyDefinitionType {
 }
 
 impl MethodDefinitionKind {
+    #[inline]
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast_impl/js.rs:1384]: Again, do in a separate PR please.

File: crates/oxc_ast/src/ast_impl/js.rs [hunk 12]
```diff
@@ -1306,28 +1381,32 @@ impl PropertyDefinitionType {
...
     pub fn is_get(&self) -> bool {
         matches!(self, Self::Get)
     }
 
-    pub fn scope_flags(self) -> ScopeFlags {
+    pub const fn scope_flags(self) -> ScopeFlags {
         match self {
-            Self::Constructor => ScopeFlags::Constructor | ScopeFlags::Function,
+            Self::Constructor => ScopeFlags::Constructor.union(ScopeFlags::Function),
```
Comment (overlookmotel) [inline crates/oxc_ast/src/ast_impl/js.rs:1406]: 1. As with above comments, this should be a separate PR.
2. Personally I prefer it as it was before. I find ` | ` clearer than `.union()`.
3. What is the rationale for making this function `const`?

**Comment by overlookmotel** [comment]
Thanks *loads* for doing this Don. I in particular will really appreciate the comments on the TS types, because I am not a habitual TS user, and so I'm constantly wracking my brains "but what *is* a TSIndexSignature?"

I've made a lot of comments. I hope this comes across in the spirit in which it's intended - helping spot a few small things, rather than gratuitous nit-picking. You can feel free to ignore the stylistic suggestions if you don't have time.

The main thing, though, is to please make this PR just the comments, and separate out the addition of various impls + methods into separate PRs. Some of those changes may be more questionable, and it'd be helpful to see them and discuss them in isolation. You are set up on Graphite now, right?

**Comment by DonIsaac** [comment]
Yeah, I've got graphite set up. I'll create a stack with the other impls on a downstream PR

**Comment by overlookmotel** [comment]
There's still some methods added in this PR e.g. `impl<'a> Deref for BindingIdentifier<'a>`. As discussed above, please move them into a separate PR.

**Comment by DonIsaac** [comment]
@overlookmotel all `Deref` impls and new methods have been removed.

**Comment by DonIsaac** [comment]
@overlookmotel bumping this

**Comment by Boshen** [comment]
This conflicts with https://github.com/oxc-project/oxc/pull/4051

Shall we try and consolidate both? Merging this first is fine.
